### PR TITLE
feat: compile and push ampd contracts to CF r2 bucket

### PR DESCRIPTION
**Pull request by maancham**
## Description

[AXE-3646]

## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

## Other Notes

[AXE-3646]: https://axelarnetwork.atlassian.net/browse/AXE-3646?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/build-ampd-and-push-to-r2.yaml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+          echo "wasm-directory=./artifacts" >> $GITHUB_OUTPUT
+
+      - uses: ryand56/r2-upload-action@latest
+        with:
+          r2-account-id: ${{ secrets.R2_ACCOUNT_ID }}
+          r2-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID_CF }}
+          r2-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY_CF }}
+          r2-bucket: ${{ secrets.R2_BUCKET }}
+          source-dir: ${{ steps.compile-contracts.outputs.wasm-directory }}
+          destination-dir: ./pre-releases/ampd/
```
Comment (talalashraf) [inline .github/workflows/build-ampd-and-push-to-r2.yaml:61]: add /contracts to destination dir

File: .github/workflows/build-ampd-and-push-to-r2.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+          fetch-depth: "0"
+          path: axelar-amplifier
+          submodules: recursive
+          ref: ${{ steps.get-branch-name.outputs.branch }}
+
+      - name: Compile amplifier contracts
+        id: compile-contracts
+        run: |
+          cd axelar-amplifier
+          docker run --rm -v "$(pwd)":/code --mount type=volume,source="$(basename "$(pwd)")_cache",target=/code/target --mount type=volume,source=registry_cache,target=/usr/local/cargo/registry cosmwasm/rust-optimizer:0.15.1
```
Comment (talalashraf) [inline .github/workflows/build-ampd-and-push-to-r2.yaml:47]: any specific reason this is our approach as opposed to cargo build in a linux machine ?

**Review by talalashraf — APPROVED** [review]

File: .github/workflows/build-ampd-and-push-to-r2.yaml (hunk 2 above)
Comment (maancham) [reply .github/workflows/build-ampd-and-push-to-r2.yaml:47]: following existing amplifier compilation method in depoloy-devnet files:
https://github.com/axelarnetwork/axelarate/blob/main/.github/workflows/reuseable-deploy-devnet.yaml#L222
Also, the use of `rust-optimizer` that `cargo build` does not handle on itself I guess.

Comment (maancham) [reply .github/workflows/build-ampd-and-push-to-r2.yaml:47]: @cgorenflo Can you also comment please?

File: .github/workflows/build-ampd-and-push-to-r2.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+          echo "wasm-directory=./artifacts" >> $GITHUB_OUTPUT
+
+      - uses: ryand56/r2-upload-action@latest
+        with:
+          r2-account-id: ${{ secrets.R2_ACCOUNT_ID }}
+          r2-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID_CF }}
+          r2-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY_CF }}
+          r2-bucket: ${{ secrets.R2_BUCKET }}
+          source-dir: ${{ steps.compile-contracts.outputs.wasm-directory }}
+          destination-dir: ./pre-releases/ampd/contracts/
```
Comment (cgorenflo) [inline .github/workflows/build-ampd-and-push-to-r2.yaml:61]:
```suggestion
          destination-dir: ./pre-releases/ampd/contracts/

```

File: .github/workflows/build-ampd-and-push-to-r2.yaml (hunk 2 above)
Comment (cgorenflo) [reply .github/workflows/build-ampd-and-push-to-r2.yaml:47]: the `cosmwasm/rust-optimizer` image compiles and optimizes the contract binaries in a reproducible way. It's the recommended approach for building cosmwasm contracts

**Review by cgorenflo — APPROVED** [review]

**Pull request merged**